From 42b54c78233d8db2e4cb537fc9d1fda066e53522 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Mon, 10 Feb 2025 13:36:32 +0800
Subject: [PATCH 087/104] mtk: mt76: mt7996: fix antenna mask config when path
 is larger than then nss

The maximum path and nss may be different so add orig_antenna_mask to
prevent setting invalid nss.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt7996/eeprom.c | 1 +
 mt7996/main.c   | 3 ++-
 mt7996/mt7996.h | 1 +
 3 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/mt7996/eeprom.c b/mt7996/eeprom.c
index 05450ba9..079fb4b3 100644
--- a/mt7996/eeprom.c
+++ b/mt7996/eeprom.c
@@ -474,6 +474,7 @@ int mt7996_eeprom_parse_hw_cap(struct mt7996_dev *dev, struct mt7996_phy *phy)
 		phy->has_aux_rx = true;
 
 	mphy->antenna_mask = BIT(nss) - 1;
+	phy->orig_antenna_mask = mphy->antenna_mask;
 	mphy->chainmask = (BIT(path) - 1) << dev->chainshift[band_idx];
 	phy->orig_chainmask = mphy->chainmask;
 	dev->chainmask |= mphy->chainmask;
diff --git a/mt7996/main.c b/mt7996/main.c
index c645a625..536695fd 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -1881,7 +1881,8 @@ mt7996_set_antenna(struct ieee80211_hw *hw, u32 tx_ant, u32 rx_ant)
 		u8 shift = dev->chainshift[band_idx];
 
 		phy->mt76->chainmask = tx_ant & phy->orig_chainmask;
-		phy->mt76->antenna_mask = phy->mt76->chainmask >> shift;
+		phy->mt76->antenna_mask = (phy->mt76->chainmask >> shift) &
+					  phy->orig_antenna_mask;
 
 		mt76_set_stream_caps(phy->mt76, true);
 		mt7996_set_stream_vht_txbf_caps(phy);
diff --git a/mt7996/mt7996.h b/mt7996/mt7996.h
index 3ec87425..ac380337 100644
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -729,6 +729,7 @@ struct mt7996_phy {
 	struct delayed_work ipi_work;
 
 	u16 orig_chainmask;
+	u16 orig_antenna_mask;
 
 	bool has_aux_rx;
 	bool counter_reset;
-- 
2.45.2

